src/analysis/processing/qgsalgorithmdownloadvectortiles.cpp

```diff
@@ -0,0 +1,225 @@
+  addParameter( new QgsProcessingParameterExtent( QStringLiteral( "EXTENT" ), QObject::tr( "Extent" ) ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "MAX_ZOOM" ), QObject::tr( "Maximum zoom level to download" ), QgsProcessingParameterNumber::Integer, 14, false, 0 ) );
+
+  std::unique_ptr< QgsProcessingParameterNumber > tileLimitParam = std::make_unique< QgsProcessingParameterNumber >( QStringLiteral( "TILE_LIMIT" ), QObject::tr( "Tile limit" ), QgsProcessingParameterNumber::Integer, 10000, false, 0 );
+  tileLimitParam->setFlags( tileLimitParam->flags() | QgsProcessingParameterDefinition::FlagAdvanced );
+  addParameter( tileLimitParam.release() );
+
+  addParameter( new QgsProcessingParameterVectorTileDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Output" ) ) );
+}
+
+QVariantMap QgsDownloadVectorTilesAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QgsMapLayer *layer = parameterAsLayer( parameters, QStringLiteral( "INPUT" ), context );
+  if ( !layer )
+    throw QgsProcessingException( QObject::tr( "Invalid input layer" ) );
+
+  const QgsRectangle extent = parameterAsExtent( parameters, QStringLiteral( "EXTENT" ), context, layer->crs() );
+  int maxZoomLevel = parameterAsInt( parameters, QStringLiteral( "MAX_ZOOM" ), context );
+  const int tileLimit = parameterAsInt( parameters, QStringLiteral( "TILE_LIMIT" ), context );
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+
+  QgsVectorTileLayer *vtLayer = qobject_cast< QgsVectorTileLayer * >( layer );
+  const QgsVectorTileDataProvider *provider = qgis::down_cast< const QgsVectorTileDataProvider * >( vtLayer->dataProvider() );
+  QgsVectorTileMatrixSet tileMatrixSet = vtLayer->tileMatrixSet();
+
+  if ( maxZoomLevel > vtLayer->sourceMaxZoom() )
+  {
+    throw QgsProcessingException( QObject::tr( "Requested maximum zoom level is bigger than available zoom level in the source layer. Please, select zoom level lower or equal to %1." ).arg( vtLayer->sourceMaxZoom() ) );
+  }
+
+  // count total number of tiles in the requested extent and zoom levels to see if it exceeds the tile limit
+  long long tileCount = 0;
+  QMap<int, QgsTileRange> tileRanges;
+  for ( int i = 0; i <= maxZoomLevel; i++ )
+  {
+    QgsTileMatrix tileMatrix = tileMatrixSet.tileMatrix( i );
+    QgsTileRange tileRange = tileMatrix.tileRangeFromExtent( extent );
+    tileRanges.insert( i, tileRange );
+    tileCount += ( tileRange.endColumn() - tileRange.startColumn() + 1 ) * ( tileRange.endRow() - tileRange.startRow() + 1 );
+    if ( tileCount > tileLimit )
+    {
+      throw QgsProcessingException( QObject::tr( "Requested number of tiles exceeds limit of %1 tiles. Please, select a smaller extent, reduce maximum zoom level or increase tile limit." ).arg( tileLimit ) );
+    }
```

Comment:
How about we move this check to outside of the zoom level loop, and then include the actual number of required tiles in the error message too? That'd save a guessing game for users who would otherwise need to iteratively increase the limit until it's sufficient for their extent.

Reply:
Fixed, thanks for the suggestion.